server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -1591,39 +1593,40 @@ private IndexMetadata convertLegacyIndex(IndexMetadata snapshotIndexMetadata, Cl
         if (snapshotIndexMetadata.getCreationVersion().before(Version.fromString("5.0.0"))) {
             throw new IllegalArgumentException("can't restore an index created before version 5.0.0");
         }
+        IndexMetadata.Builder convertedIndexMetadata = IndexMetadata.builder(snapshotIndexMetadata);
         MappingMetadata mappingMetadata = snapshotIndexMetadata.mapping();
-        Map<String, Object> loadedMappingSource = mappingMetadata.rawSourceAsMap();
-
-        // store old mapping under _meta/legacy_mappings
-        Map<String, Object> legacyMapping = new LinkedHashMap<>();
-        boolean sourceOnlySnapshot = snapshotIndexMetadata.getSettings().getAsBoolean("index.source_only", false);
-        if (sourceOnlySnapshot) {
-            // actual mapping is under "_meta" (but strip type first)
-            Object sourceOnlyMeta = mappingMetadata.sourceAsMap().get("_meta");
-            if (sourceOnlyMeta instanceof Map<?, ?> sourceOnlyMetaMap) {
-                legacyMapping.put("legacy_mappings", sourceOnlyMetaMap);
+        if (mappingMetadata != null) {
```

Comment:
the newly added test uncovered a bug as mappings can be null sometimes (even if not very useful, it's still a thing). This method was just refactored to put the existing logic under an if != null